File: python/UnionFind.py
# Union-Find Tree
class UnionFind:
    def __init__(self, n):
        self.n = n
        self.data = [0]*n
        for i in xrange(n):
            self.data[i] = i
    def root(self, x):
        if self.data[x] == x:
            return x
        self.data[x] = self.root(self.data[x])
        return self.data[x]
    def unite(self, x, y):
        rx = self.root(x)
        ry = self.root(y)
        if rx<ry:
            self.data[ry] = rx
        elif rx>ry:
            self.data[rx] = ry
    def match(self, x, y):
        return self.root(x)==self.root(y)

File: cpp/union_find/union_find.cpp
#define N 100008

int parent[N];
int root(int x) {
  if(x == parent[x]) {
    return x;
  }
  return parent[x] = root(parent[x]);
}

bool unite(int x, int y) {
  int px = root(x), py = root(y);
  if(px == py) return false;

  if(px < py) {
    parent[py] = px;
  } else {
    parent[px] = py;
  }
  return true;
}

// 以下で初期化
//
// rep(i, n) parent[i] = i;
//